pkg/scheduler/core/division_algorithm.go

```diff
@@ -20,6 +20,82 @@ func (a TargetClustersList) Len() int           { return len(a) }
 func (a TargetClustersList) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
 func (a TargetClustersList) Less(i, j int) bool { return a[i].Replicas > a[j].Replicas }
 
+type dispenser struct {
+	numReplicas int32
+	result      []workv1alpha2.TargetCluster
+}
+
+func newDispenser(numReplicas int32, init []workv1alpha2.TargetCluster) *dispenser {
+	cp := make([]workv1alpha2.TargetCluster, len(init))
+	copy(cp, init)
+	return &dispenser{numReplicas: numReplicas, result: cp}
+}
+
+func (a *dispenser) done() bool {
+	return a.numReplicas == 0 && len(a.result) != 0
+}
+
+func (a *dispenser) takeByWeight(w helper.ClusterWeightInfoList) {
+	if a.done() {
+		return
+	}
+	sum := w.GetWeightSum()
+	if sum == 0 {
+		return
+	}
+
+	sort.Sort(w)
+
+	result := make([]workv1alpha2.TargetCluster, 0, w.Len())
+	remain := a.numReplicas
+	for _, info := range w {
+		replicas := int32(info.Weight * int64(a.numReplicas) / sum)
+		result = append(result, workv1alpha2.TargetCluster{
+			Name:     info.ClusterName,
+			Replicas: replicas,
+		})
+		remain -= replicas
+	}
+	// TODO(Garrybest): take rest replicas by fraction part
+	for i := range result {
+		if remain == 0 {
+			break
+		}
+		result[i].Replicas++
+		remain--
+	}
+
+	a.numReplicas = remain
+	a.result = util.MergeTargetClusters(a.result, result)
+}
+
+func getStaticWeightInfoList(clusters []*clusterv1alpha1.Cluster, weightList []policyv1alpha1.StaticClusterWeight) helper.ClusterWeightInfoList {
+	list := make(helper.ClusterWeightInfoList, 0)
+	for _, cluster := range clusters {
+		var weight int64
+		for _, staticWeightRule := range weightList {
+			if util.ClusterMatches(cluster, staticWeightRule.TargetCluster) {
+				weight += staticWeightRule.Weight
```

Comment:
What if a cluster matches by more than one `weightRule`? 
Should we get the maximum weight instead of summing it?

Reply:
Good idea! Sum may be a confusion.